### PR TITLE
Enable ClojureScript REPL and hot reloading of CLJS code

### LINKED ISSUE #32674: Get the CLJS REPL connecting to Metabase running in the browser
### Describe the bug

When working in the CLJS environment, the Node REPL runs in its own world, separate from the browser. Some functionality doesn't work at all there (the DOM for example), and other things are different from the browser environment.

One notable gap is that `metabase.lib.core` doesn't work at all in the Node environment, but works fine in the browser. Without that "kitchen sink" file, `defmethod`s in various `metabase.lib.*` namespaces may not be loaded, which causes odd behavior when trying to debug CLJS-only issues.

In addition, connecting to the live, running browser instance is very powerful because it gives access to the current real metadata, the current query, and more.

### To Reproduce

`yarn cljs-nrepl` starts an impoverished NodeJS environment with an nrepl connection.

### Expected behavior

This (or a similar command) should connect to the live Metabase instance in your browser.

### Logs

_No response_

### Information about your Metabase installation

```JSON
master
```

### Severity

Annoying when debugging CLJS-only issues; FE integration for MLv2.

### Additional context

This doesn't work out of the box because of our nonstandard `shadow-cljs` configuration.

The typical pattern `shadow-cljs` expects for a browser app is to combine some JS dependencies, CLJS dependencies, and our own CLJS code, and compile it. It expects that this output is what ships to the browser. In dev mode, extra libraries and startup code support connecting a Clojure REPL to the browser environment.

Our configuration is to compile our CLJS code into a module, which webpack will combine with the FE code.

There's no reason that we can't include the necessary libraries and startup code in this process, it just doesn't work out of the box and isn't trivial to include.

### PR DESCRIPTION
**Pull request by bshepherdson**
Requiring the shadow-cljs browser REPL components in
metabase.util.devtools is all that was needed to get the REPL working.

But that exposed a problem - the REPL kept disconnecting when edits were
made to CLJS files. The webpack HotModuleReload plugin does not know
how to handle these changes, so it falls back to the safe thing: full
reload.

This change enables shadow-cljs hot reload, excludes the CLJS build from webpack's HMR, and enables the CLJS REPL.

## Output location

This also moves the CLJS output out of `frontend/src/cljs{,_release}` to
`target/cljs_{dev,release}`. If that causes problems for FE tooling I
don't know about, it could be moved back.

## Hot reloading

CLJS changes are hot-reloaded into the running app by Shadow CLJS - no
action is needed. Note that the UI will not refresh; you usually have to
click around so the updated MLv2 code gets executed.

## Usage of the CLJS REPL

- Open the browser (you should see `shadow-cljs: #N ready!` in console)
- Connect to nrepl on 50655
- "Select" the `:app` build;
    - https://shadow-cljs.github.io/docs/UsersGuide.html#_editor_integration
    - Conjure: `:ConjureShadowSelect app`
    - Plain REPL: Eval in CLJ mode `(shadow/repl :app)`
- Consider running `yarn build-hot:cljs` and `yarn build-hot:js` in
  separate terminals; then when `yarn build-hot:js` crashes you don't
  have to reconnect your REPL.

Note that requiring files not already in the build (eg. tests) isn't likely to
work. That might be fixable but it doesn't work right now.

## "Stale" error in browser

- This can happen if the webpack bundle is created before the CLJS is
  done building.
- If you ran `yarn build-hot:cljs` and `:js` separately, let `:cljs`
  finish compiling first.
- If you see this otherwise: (a) contact Braden; (b) either restart
  `build-hot:js` or force webpack to rebuild by hitting
  http://localhost:8080/webpack-dev-server/invalidate

## Yarn commands

`dev`, `dev-ee`, `build-hot`, `build-static-viz:watch` all now wait for
the CLJS build to be complete by watching for files to be generated.
They should work the same from a user's POV.

(Since the webpack build no longer watches the CLJS output,
it won't notice the changes and immediately rebuild after the first
build. This avoids errors on the clean build, or "stale build" messages
in the app from Shadow CLJS.)

Fixes https://github.com/metabase/metabase/issues/32674.

### CODE REVIEW

File: src/metabase/lib/js.cljs [hunk 1]
```diff
@@ -873,3 +873,34 @@
    Can be passed an integer table id or a legacy `card__<id>` string."
   [a-query table-id]
   (lib.core/with-different-table a-query table-id))
+
+;; This hooks into Webpack's Hot Module Reload, only in dev mode, to accept HMR when the CLJS changes.
+;; Shadow-cljs rebuilds the files, and webpack sees the change and reloads a subtree up to this file. Since there's no
+;; state held anywhere here, just library, code, there's no further action needed.
```
Comment (metamben) [inline src/metabase/lib/js.cljs:879]:
```suggestion
;; state held anywhere here, just library code, there's no further action needed.
```

File: src/metabase/lib/js.cljs [hunk 2]
```diff
@@ -873,3 +873,34 @@
...
+;; TODO: shadow-cljs can update the code more swiftly when running in dev mode, but there are two problems.
+;; - The webpack build changes goog.provide/goog.require into __webpack_require__ etc., but shadow-cljs does not.
+;;   Therefore the code loaded by shadow-cljs won't compile in the browser.
+;; - Even if it did work, I could find a way to tell webpack's HMR to allow a hot reload (ie. don't refresh the page)
```
Comment (metamben) [inline src/metabase/lib/js.cljs:883]:
```suggestion
;; - Even if it did work, I could find no way to tell webpack's HMR to allow a hot reload (ie. don't refresh the page)
```

**Review by metamben — COMMENTED** [review]
It's an interesting read, someone with more knowledge should review.

**Comment by shaunlebron** [comment]
Awesome, great stuff!

For the work remaining:

1. I’m not sure what is being missed in the whole App re-render, but maybe we can find some way to call the top-level init, which will re-render App with the updated reducers and routes: https://github.com/metabase/metabase/blob/76b54609e5f8f51857e354f2dda58f70ab564e47/frontend/src/metabase/app.js#L65-L89

2. I think we can do a `:shadow.build/stage :compile-prepare` hook that:
   * only runs if `:shadow.build/mode` is `:dev`
   * gets the sources in `(:sources build-state)` associated with `(:entries (:config build-state))`
   * return `(update-in build-state [:sources entry :source] add-hmr)` for each `entry`, where `add-hmr` is a function which appends the hrm cljs code to the source string

**Review by shaunlebron — CHANGES_REQUESTED** [review]
Awesome, great stuff!

For the work remaining:

1. I’m not sure what is being missed in the whole App re-render, but maybe we can find some way to call the top-level init, which will re-render App with the updated reducers and routes: https://github.com/metabase/metabase/blob/76b54609e5f8f51857e354f2dda58f70ab564e47/frontend/src/metabase/app.js#L65-L89

2. I think we can do a `:shadow.build/stage :compile-prepare` hook that:
   * only runs if `:shadow.build/mode` is `:dev`
   * gets the sources in `(:sources build-state)` associated with `(:entries (:config build-state))`
   * return `(update-in build-state [:sources entry :source] add-hmr)` for each `entry`, where `add-hmr` is a function which appends the hrm cljs code to the source string

**Review by metamben — APPROVED** [review]
Nice!

**Review by shaunlebron — APPROVED** [review]
Really nice job on this.  I tried it out in Conjure, and it worked well and verified that hot reloading still works on JS files. 🎉

**Pull request merged**